File: src/main/java/com/gmail/gregorysalsbery/treevolution/environment/Sun.java
package com.gmail.gregorysalsbery.treevolution.environment;

import com.gmail.gregorysalsbery.treevolution.grid.GridPoint;
import com.gmail.gregorysalsbery.treevolution.tree.Tree;
import com.gmail.gregorysalsbery.treevolution.tree.parts.TreePart;
import com.gmail.gregorysalsbery.treevolution.util.Config;
import lombok.Setter;

import java.util.List;

public class Sun {

    // TODO: need to find a more efficient method for doing the sun
    // the current method literally looks at every space and then looks at every individual tree part for each space
    // maybe each treepart could register its position to a new Grid class
    // the grid class could maintain lists of everything in a particular column
    // tree parts never change their x positions so the lists per column would only have to change to add new growths
    // then we can sort the column lists by y position and assign sunlight down the list
    // should be much faster!

    private static final int START_LIGHT_LEVEL = 10;

    @Setter
    private List<Tree> trees;

    public Sun(List<Tree> trees) {
        this.trees = trees;
    }

    public void shine() {
        for(int x=0; x<Config.GRID_SIZE_X; x++) {
            shineColumn(x);
        }
    }

    public void shineColumn(int x) {
        int lightLevel = START_LIGHT_LEVEL;
        for(int y=Config.GRID_SIZE_Y-1; y>=0; y--) {
            lightLevel = shineSpace(x, y, lightLevel);
            // if all light is used up for the column, then exit the loop for that column
            if(lightLevel <= 0) break;
        }
    }

    public int shineSpace(int x, int y, int lightLevel) {
        for(Tree tree : trees) {
            for(TreePart treePart : tree.getTreeParts()) {
                if(treePart.getXy().compare(x, y)) {
                    // assume only one tree part can occupy a space, no overlaps! (will have to actually make that happen somehow)
                    return treePart.hitByLight(lightLevel);
                }
            }
        }
        // no tree parts in the space
        return lightLevel;
    }
}

File: environment/sun.py
import config


class LightColumn:
    def __init__(self):
        self.column_of_tree_parts = list()

    def register_part(self, tree_part):
        self.column_of_tree_parts.append(tree_part)

    def shine(self, light_level):
        # sort the column in reverse order so that the highest y is first
        self.column_of_tree_parts.sort(key=lambda p: p.get_y(), reverse=True)

        # iterate down the column, applying light level appropriately
        for part in self.column_of_tree_parts:
            # if two parts happen to occupy the same space, then whichever happens to be first will get the most
            # light, the next one getting the reduced light
            light_level = part.hit_by_light(light_level)
            # no need to iterate over the tree parts below if the light level has been reduced to 0
            if light_level <= 0:
                break


class Sun:
    START_LIGHT_LEVEL = 10

    def __init__(self):
        self.light_columns = dict()
        # add a light column for each x position in the grid
        for x in range(0, config.GRID_SIZE_X):
            self.light_columns[x] = LightColumn()

    def register_part(self, tree_part):
        # add a tree part to its appropriate light column by x position
        self.light_columns[tree_part.get_x()].register_part(tree_part)

    def shine(self):
        # start shine for each light column
        for x in range(0, config.GRID_SIZE_X):
            self.light_columns[x].shine(Sun.START_LIGHT_LEVEL)